subsystems/viamserver/viamserver.go

```diff
@@ -294,17 +296,79 @@ func (s *viamServer) HealthCheck(ctx context.Context) (errRet error) {
 	return errRet
 }
 
+func (s *viamServer) isRestartAllowed(ctx context.Context) (restartAllowed bool, errRet error) {
+	for _, url := range []string{s.checkURL, s.checkURLAlt} {
+		s.logger.Debugf("starting restart allowed check for %s using %s", SubsysName, url)
+
+		ctx, cancelFunc := context.WithTimeout(ctx, time.Second*10)
+		defer cancelFunc()
+
+		url += "/restart_status"
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+		if err != nil {
+			return false, errw.Wrapf(err, "checking whether %s allows restart", SubsysName)
+		}
+
+		// disabling the cert verification because it doesn't work in offline mode (when connecting to localhost)
+		//nolint:gosec
+		client := &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}}
+
+		resp, err := client.Do(req)
+		if err != nil {
+			return false, errw.Wrapf(err, "checking whether %s allows restart", SubsysName)
+		}
+
+		defer func() {
+			errRet = errors.Join(errRet, resp.Body.Close())
+		}()
+
+		var restartStatusResponse rdkweb.RestartStatusResponse
+		err = json.NewDecoder(resp.Body).Decode(&restartStatusResponse)
+		if err != nil {
+			// Only debug-log errors from JSON decoding. Interacting with older
+			// viam-server instances will result in a decoding error, as empty HTML
+			// will be returned.
+			s.logger.Debugf("ignoring error from restart allowed check for %s: %v",
+				SubsysName, err.Error())
+		} else {
+			return restartStatusResponse.RestartAllowed, nil
+		}
+	}
+	return false, nil
+}
+
 func (s *viamServer) Update(ctx context.Context, cfg *pb.DeviceSubsystemConfig, newVersion bool) (bool, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	setFastStart(cfg)
+
+	// By default, return false on the needRestart flag, as we await the user to
+	// kill/restart viam-server directly.
+	var needRestart bool
+
 	if newVersion && s.running {
-		s.logger.Info("awaiting user restart to run new viam-server version")
 		s.shouldRun = false
+
+		// viam-server can be safely restarted even while running if the process
+		// has reported it is safe to do so through its `restart_status` HTTP
+		// endpoint.
+		restartAllowed, err := s.isRestartAllowed(ctx)
```

Comment:
Discussed with both of you offline about the only "risk" from the scope:
> Risk: A single config update triggers both a viam-server version update and a reconfiguration, and viam-agent tries to restart viam-server while it is in the middle of a reconfiguration due to a race.
Mitigation: (Avoid) Check the restart_status URL for a {“restart_allowed”: true} field in agent multiple times (4 over 60 seconds) in viam-agent before actually restarting viam-server.

What I have now does _not_ avoid the risk and does _not_ implement the "Mitigation" described above. 

@Otterverse I know we spoke about checking the `restart_status` endpoint once when an update is possible for `viam-server`, waiting N minutes, checking the `restart_status` endpoint again, and, only if both checks had indicated a possible external restart, restarting `viam-server`.

After discussing more offline and in the `rdk` PR with @dgottlieb, we would be in favor of the "dumber" strategy here: check the `restart_status` once when a restart is availabe and restart `viam-server` if it indicates an allowed restart.

Let's call an accidental restart of the `viam-server` while it's reconfiguring a "false positive" and any delay in restarting the `viam-server` when a new version is available and the `viam-server` is not reconfiguring a "false negative." The "N minute" strategy seems to lower false positives and raise false negatives. The "dumber" strategy I have here seems to raise false positives and lower false negatives. I would argue that neither approach eliminates false positives, as I can come up with an antagonistic case where the "N minute" strategy still results in a false positive.

If we _really_ wanted to avoid false positives, we think we would need a scope change and a slightly different protocol here. Something like a) `viam-server` handles its own shutdown for restart or b) `viam-server` returns a duration of time that it can be safely be restarted in `restart_status`. In our opinion, the false positive case is not dangerous enough to warrant that change + addition of complexity.

Let me know what you think; this is much more your codebase than NetCode's. And, thanks for all the help in implementing this.

Reply:
So I think it's much more likely to hit false positives without a delay here. When someone clicks "save" in the config page, if they made any changes that would require a reconfig and change the viam-server version, both these things kick off at the same time.

That said, if you feel it's not "dangerous" (it will be a signaled/controlled shutdown via SIGTERM after all) to hit those false positives occasionally, then that's fine by me. Simpler is nice in a lot of ways.

Reply:
> That said, if you feel it's not "dangerous" (it will be a signaled/controlled shutdown via SIGTERM after all) to hit those false positives occasionally, then that's fine by me. Simpler is nice in a lot of ways.

Yep that's the general sentiment! Will stick with the simpler method since it's alright with you.